### DISCUSSION TITLE
Allow feature registration in standalone components.

### DISCUSSION THREAD

**Discussion by BenGrn** [ngrx/platform#4288] [Ideas]
In existing module-based libraries, features can be registered with a simple `StoreModule.forFeature(booksFeature)` and `EffectsModule.forFeature(booksEffects)` in the imports array. This allows any application using this module/library to have that store active.
In standalone applications and components, the following method is provided from within the application bootstrap: `provideState(booksFeature)`.
This is an environmental provider and cannot be used within standalone components. We need a `provideFeature(booksFeature)` and a `provideFeatureEffects(booksEffects)` to allow standalone components to add store features, so libraries that require stores for use don't have an additional requirement of adding required features and effects to the application bootstrap. It's an additional step that could easily be missed and could add copious amounts of boiler plate code that should be added from the library implicitly.

**Comment by brandonroberts** [comment]
Thanks @BenGrn. We've covered why this isn't a viable option

https://github.com/ngrx/platform/issues/4234#issuecomment-1930549625
https://github.com/ngrx/platform/issues/4063#issuecomment-1761547428

**Reply by BenGrn** [reply]
Those comments address using provideState in standalone components, not creating a new provider to provide feature as such. It seems like a regression as we move away from module based applications and lose that feature. I have zero understanding of how the store is built under the hood but only want to ask if there is not a way to add/inject additional stores and effects modules after initial store creation within the bootstrap?

**Reply by brandonroberts** [reply]
I see. Yes, you can still register Store/Effects features through the `providers` array in the route config.

```ts
{
   path: 'some-feature',
   providers: [
    provideState(featureA),
    provideEffects(FeatureEffectsA, FeatureEffectsB)
  ]
}
```

If you want to register them outside the bootstrap and route providers, you have to do it manually.

**Reply by BenGrn** [reply]
Is there any documentation on how to provide these manually as the standalone components in our applications are not managed by a routers at this stage?

**Reply by brandonroberts** [reply]
There's two APIs to do these

For Store features - https://ngrx.io/api/store/ReducerManager
For Effects features - https://ngrx.io/api/effects/EffectSources (You will need to instantiate the Effect class first)

You can also look at the source code for `provideState` and `provideEffects` because that's what they use internally

**Reply by BenGrn** [reply]
I've had a play around with those APIs for the last few hours and there just isn't enough context, or documentation to implement these manually without a thorough understanding of how NGRX works as a whole. It's a shame we have effectively lost a feature moving to standalone components that is likely going to force us to continue using ngModule based applications or look at managing state in a different way.

**Reply by brandonroberts** [reply]
I'll see if I can put together a simpler example. We brought this up with the Angular team also, but it wasn't agreed upon that some sort of initializer is needed at the standalone component level.

We technically don't have an API in Angular to do this unless you create an NgModule, use your Store/Effects features into it, and import that into your standalone component. Not a great experience, but a workaround.

**Reply by BenGrn** [reply]
Working around like this we get a `NullInjectorError` The Application would need to be a module-based application for this and not standalone?

We cant import `StoreModule.forRoot({}),` within the module as this provides two stores, adding `provideStore()` to the module's providers does not work because it is an `EnvironmentProviders`
